File: web/src/types/news.ts
/**
 * News & Sentiment shared types (mirrors `web/src/api/news.ts`).
 *
 * These types are the single source of truth for any component that
 * consumes the news / sentiment endpoints. Keep them in sync with the
 * backend Pydantic schemas in `app/schemas/news*.py`.
 */

/** Market segment identifier.
 *
 *  ``global`` is a frontend-only sentinel introduced in M22-2 (2026-07-04)
 *  to represent "all markets combined" — it is mapped on the backend
 *  to a union of the concrete markets (``cn_a``, ``us``, ``crypto`` and
 *  any legacy bucket the collector has ever written).
 */
export type NewsMarket = 'cn_a' | 'us' | 'crypto' | 'global';

/** Sentiment label output by the LLM processor. */
export type SentimentLabel = 'negative' | 'neutral' | 'positive';

/** Importance bucket (1 = informational, 5 = market-moving). */
export type ImportanceLevel = 1 | 2 | 3 | 4 | 5;

/** A symbol mentioned by a news article. */
export interface NewsSymbol {
  symbol: string;
  market?: string | null;
  /** How the symbol was matched (e.g. ``ticker``, ``name``, ``alias``). */
  match_type?: string | null;
  /** Instrument display name (cached from ``etf_info``). */
  name?: string | null;
  /** Chinese display name (cached from ``etf_info``). */
  name_zh?: string | null;
}

/** Per-article engagement metrics (likes / shares / …). */
export interface NewsEngagement {
  likes?: number;
  comments?: number;
  shares?: number;
  views?: number;
  [key: string]: number | undefined;
}

/** A single news article surfaced by the news collector. */
export interface NewsArticle {
  id: number;
  /** Source identifier, e.g. ``xinhua``, ``xueqiu``, ``reddit``… */
  source: string;
  url: string;
  market: NewsMarket;
  language: string;
  title: string;
  body: string | null;
  author: string | null;
  published_at: string;
  fetched_at: string;
  engagement: NewsEngagement;
  /** Normalized score in ``[-1, 1]`` if processed by the LLM. */
  sentiment_score: number | null;
  sentiment_label: SentimentLabel | null;
  /** Confidence of the LLM sentiment call, in ``[0, 1]``. */
  sentiment_confidence: number | null;
  /** Optional list of driver phrases extracted by the LLM. */
  sentiment_drivers: string[] | null;
  event_category: string | null;
  importance: ImportanceLevel | null;
  symbols: NewsSymbol[];
  /** Lazily-fetched full body (Jina Reader cache). ``null`` when never loaded. */
  full_content: string | null;
  /** ISO timestamp of the last successful Jina fetch (cache TTL anchor). */
  full_content_fetched_at: string | null;
  /** Cached Chinese translation (DeepSeek). Populated only for English articles. */
  translated_zh: string | null;
  /** ISO timestamp of the last successful DeepSeek translation. */
  translation_generated_at: string | null;
  /**
   * AI-cleanup observability (M22-3, 2026-07-05).
   *
   * - ``"cleaned"``  — DeepSeek call succeeded; ``full_content`` is the cleaned body.
   * - ``"skipped"``  — DeepSeek was unconfigured; ``full_content`` is raw Jina Markdown.
   * - ``"failed"``   — DeepSeek raised; ``full_content`` is raw Jina Markdown.
   * - ``"not_attempted"`` — reserved for a future batch path.
   * - ``null``       — scheduler never reached this article. The detail page
   *                    renders the yellow "尚未抓取正文" banner in this case.
   */
  ai_cleanup_status: 'cleaned' | 'skipped' | 'failed' | 'not_attempted' | null;
  /** ISO timestamp of the last AI-cleanup attempt (``null`` if never tried). */
  ai_cleaned_at: string | null;
}

/** Response shape for ``POST /news/{id}/fetch-content``. */
export interface NewsFetchContentResponse {
  success: boolean;
  /** Cached Markdown body, or a fallback to the intro on failure. */
  content: string | null;
  cached: boolean;
  error: string | null;
}

/** Response shape for ``POST /news/{id}/translate``. */
export interface NewsTranslateResponse {
  /** The Chinese translation (Markdown). May equal the cached value. */
  translation: string;
  /** True if we returned a previously-stored row (no LLM call). */
  cached: boolean;
  /** Tokens consumed by the LLM call; null on cache hit or no-usage response. */
  tokens_used: number | null;
  /** ISO timestamp of the translation (cache anchor). */
  generated_at: string | null;
  /** Source language code as stored on the article. */
  source_language: string;
  /** Target language code, e.g. ``zh``. */
  target_language: string;
}

/** Paginated list response. */
export interface NewsListResponse {
  items: NewsArticle[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

/** List query parameters. */
export interface NewsListParams {
  market?: NewsMarket | string;
  symbol?: string;
  source?: string;
  from_date?: string;
  to_date?: string;
  /** Full-text search query (best-effort, server-side). */
  q?: string;
  page?: number;
  page_size?: number;
  importance_min?: ImportanceLevel;
  /**
   * Filter by one or more ``event_category`` values. The backend
   * accepts a list (``?event_category=geopolitics&event_category=central_bank``)
   * — repeat the query parameter to OR multiple categories together.
   *
   * Allowed values mirror the LLM prompt in
   * ``app/services/news/sentiment/prompts.py``:
   * earnings|m&a|product|macro|regulation|guidance|analyst|legal|rumor
   * |geopolitics|central_bank|election|trade_war|sanction|other
   */
  event_category?: string[];
}

/** Watchlist-scoped metadata returned by /news/watchlist. */
export interface NewsWatchlistMeta {
  /** All favorite instrument codes the user has on file. */
  symbols: string[];
  /** How many of those symbols have at least one matching article. */
  symbols_with_news: number;
  /** Total matching articles across all pages. */
  total_articles: number;
}

/** Response shape for /news/watchlist — same as ``NewsListResponse``
 *  plus a ``watchlist`` metadata block. */
export interface NewsWatchlistResponse extends NewsListResponse {
  watchlist: NewsWatchlistMeta;
}

/** Query parameters for the watchlist-scoped list endpoint.
 *  Note: ``symbol`` is intentionally absent — the symbol set comes from
 *  the user's favorites on the server side. */
export interface NewsWatchlistParams {
  market?: NewsMarket | string;
  source?: string;
  from_date?: string;
  to_date?: string;
  page?: number;
  page_size?: number;
  event_category?: string[];
}

/** Per-source count summary returned by `/news/stats/sources`. */
export interface NewsSourceStat {
  source: string;
  total: number;
  last_7d: number;
  last_24h: number;
}

/** Last etl_log row for a single news source. */
export interface NewsSourceLatestEtl {
  status: string;
  records: number | null;
  error_msg: string | null;
  finished_at: string | null;
  started_at: string | null;
}

/** Per-source diagnostics row returned by /news/health. */
export interface NewsSourceHealth {
  source: string;
  job_id: string | null;
  total: number;
  last_24h: number;
  last_published_at: string | null;
  last_fetched_at: string | null;
  latest_etl: NewsSourceLatestEtl | null;
}

/** APScheduler job snapshot entry. */
export interface NewsSchedulerJob {
  id: string;
  name: string;
  next_run_time: string | null;
}

/** Status of a single news / sentiment worker job. */
export interface NewsWorkerStatus {
  name: string;
  label: string;
  schedule: string;
  last_run: string | null;
  last_status: 'success' | 'failed' | 'never_run' | 'unknown';
  last_records: number | null;
  last_error: string | null;
  articles_24h: number;
}

/** Full response shape of /news/health. */
export interface NewsHealthResponse {
  as_of: string;
  scheduler_running: boolean;
  scheduler_jobs: NewsSchedulerJob[];
  scheduler_total_jobs: number;
  sources: NewsSourceHealth[];
  /**
   * M22-4 (2026-07-18) worker health grid — one row per news / sentiment
   * scheduler job.  Mirrors the diagnostics shown on the NewsHealth page.
   */
  workers: NewsWorkerStatus[];
  /**
   * M22-3 (2026-07-05) AI-cleanup observability — breakdown of the
   * last 24 hours of ``news_article`` rows whose
   * ``ai_cleaned_at`` is within the window.
   */
  ai_cleanup_24h: {
    /** Total rows the scheduler reached in the last 24h. */
    total: number;
    /** Rows whose AI cleanup actually ran (DeepSeek returned a body). */
    cleaned: number;
    /** Rows where DeepSeek was unconfigured (``skipped``). */
    skipped: number;
    /** Rows where DeepSeek raised or returned a too-short body. */
    failed: number;
    /** cleaned / (cleaned + failed) * 100, 1 decimal place. */
    cleaned_pct: number;
    /** Threshold below which the dashboard flips the alert card on. */
    alert_threshold_pct: number;
    /** Convenience flag — ``cleaned_pct < alert_threshold_pct``. */
    alert: boolean;
  };
}

/** A single event-driven signal returned by ``/news/event-signals``. */
export interface EventSignal {
  id: number;
  title: string;
  source: string;
  url: string;
  market: NewsMarket;
  event_category: string;
  importance: ImportanceLevel;
  sentiment_score: number | null;
  sentiment_label: SentimentLabel | null;
  published_at: string;
  summary: string | null;
  symbols: NewsSymbol[];
  signal_direction: 'bullish' | 'bearish' | 'neutral';
  signal_strength: number;
}

/** Query parameters for ``GET /news/event-signals``. */
export interface EventSignalsParams {
  days?: number;
  market?: NewsMarket | string;
  event_category?: string[];
  limit?: number;
}

/** Response shape for ``GET /news/event-signals``. */
export interface EventSignalsResponse {
  items: EventSignal[];
}

/** Theme weight in retail sentiment aggregation. */
export interface RetailTheme {
  theme: string;
  percentage: number;
}

/** Aggregated retail-sentiment output. */
export interface RetailSentiment {
  /** Overall sentiment score in ``[-1, 1]``. */
  overall: number;
  bull_bear_ratio: { bull: number; bear: number };
  main_themes: RetailTheme[];
  /** Controversy index, ``[0, 1]`` (higher = more disagreement). */
  controversy: number;
  /** LLM-generated summary of retail discussion. */
  summary: string;
}

/**
 * Per-symbol sentiment aggregate returned by ``GET /research/sentiment-data/aggregate``.
 */
export interface SentimentAggregateItem {
  symbol: string;
  /** Backend alias used by the research API; mapped to ``symbol`` if needed. */
  instrument_code?: string | null;
  name?: string | null;
  name_zh?: string | null;
  market?: string | null;
  /** Average sentiment score in ``[-1, 1]``. */
  avg_score: number;
  /** Overall sentiment label derived from ``avg_score``. */
  label: SentimentLabel;
  /** Number of contributing records. */
  count: number;
  /** Count of positive-labeled records. */
  bull: number;
  /** Count of negative-labeled records. */
  bear: number;
  /** Count of neutral-labeled records. */
  neutral: number;
  /** Daily average scores over the requested window (oldest -> newest). */
  sparkline: number[];
  /** Title of the most recent record. */
  latest_title?: string | null;
  /** ``news_article.id`` of the most recent record, when available. */
  latest_id?: number | null;
  /** Direct URL of the most recent record, when available. */
  latest_url?: string | null;
  /** ISO timestamp of the most recent record. */
  latest_published_at?: string | null;
}

/** Alias for consumers that expect the older name. */
export type EventSignalListResponse = EventSignalsResponse;

File: app/models/news.py
"""News / social-media ORM models.

Stores raw articles and posts from RSS feeds, SEC filings and Reddit,
plus a normalised many-to-many link to instrument codes. Engagement
metadata (upvotes, comments, ratios) is kept in a JSONB column so
source-specific fields do not require schema changes.

Tables:
  - news_article           : one row per article / post
  - news_article_symbol    : many-to-many ticker link
  - reddit_comment_cache   : optional second-level discussion cache
"""

from datetime import datetime

from sqlalchemy import (
    BigInteger,
    Column,
    DateTime,
    Float,
    ForeignKey,
    Index,
    Integer,
    SmallInteger,
    String,
    Text,
    UniqueConstraint,
    func,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship

from app.core.database import Base


# Use JSON type with fallback for SQLite tests.
try:  # pragma: no cover - import-time branch
    from sqlalchemy import JSON  # type: ignore
    _JSON_TYPE = JSON().with_variant(JSONB(), "postgresql")
except Exception:  # pragma: no cover
    from sqlalchemy import JSON  # type: ignore
    _JSON_TYPE = JSON()


class NewsArticle(Base):
    """A news article or social-media post.

    The combination of (source, source_id) is unique, allowing safe
    upserts and de-duplication without a UUID primary key.
    """

    __tablename__ = "news_article"

    id = Column(Integer, primary_key=True, autoincrement=True)
    source = Column(String(50), nullable=False, comment="yahoo_finance | cnbc | sec_edgar | reddit | ...")
    source_id = Column(String(200), nullable=False, comment="Source-native id (guid, sec accession, reddit id)")
    url = Column(String(1000), nullable=False)
    url_hash = Column(String(32), nullable=False, unique=True, index=True, comment="MD5(url) for dedup")
    content_hash = Column(String(32), index=True, comment="simhash for near-duplicate detection")
    title = Column(String(1000), nullable=False)
    summary = Column(Text, comment="Body excerpt or RSS summary")
    body = Column(Text, comment="Full body if available")
    body_html = Column(Text, comment="Raw HTML body for re-parsing")
    full_content = Column(Text, comment="Fetched full body via Jina Reader (cached)")
    full_content_fetched_at = Column(
        DateTime,
        comment="When the Jina Reader fetch last ran (for cache TTL)",
    )
    author = Column(String(200))
    author_followers = Column(Integer, comment="Followers/fans for retail sources (Xueqiu/Reddit)")
    language = Column(String(10), default="en")
    market = Column(String(20), default="US", comment="US / HK / CN / CRYPTO / GLOBAL")
    published_at = Column(DateTime, nullable=False, index=True)
    fetched_at = Column(DateTime, server_default=func.now(), nullable=False)
    # Optional "category" tag (top news, earnings, 8-K, options, DD, ...)
    category = Column(String(100))
    # Engagement / source-specific metrics (score, upvote_ratio, num_comments,
    # subreddit, flair, cik, form, ...). Free-form JSON.
    engagement = Column(_JSON_TYPE, comment="Source-specific engagement JSON")
    # Sentiment placeholder for Agent E. Stays NULL until the LLM pass runs.
    sentiment_score = Column(Integer, comment="-100..100 placeholder for Agent E")
    sentiment_label = Column(String(20), comment="bullish | bearish | neutral")
    sentiment_confidence = Column(Float, comment="0..1 LLM confidence")
    sentiment_drivers = Column(_JSON_TYPE, comment="List of driver keywords from LLM")
    event_category = Column(String(50), comment="earnings|m&a|product|macro|regulation|guidance|analyst|legal|rumor|other")
    importance = Column(SmallInteger, comment="1..5 LLM importance rating")
    sentiment_processed_at = Column(DateTime, comment="When LLM sentiment was filled")
    # Chinese translation cache, filled on demand by the
    # ``/news/{id}/translate`` endpoint. Populated only for English
    # articles (the API enforces ``language == 'en'`` before writing).
    translated_zh = Column(
        Text, comment="DeepSeek-generated Chinese translation of body / full_content"
    )
    translation_generated_at = Column(
        DateTime, comment="When the LLM translation was last produced"
    )

    symbols = relationship(
        "NewsArticleSymbol",
        back_populates="article",
        cascade="all, delete-orphan",
    )

    __table_args__ = (
        UniqueConstraint("source", "source_id", name="uq_news_article_source_id"),
        Index("ix_news_article_source_published", "source", "published_at"),
        Index("ix_news_article_published_desc", published_at.desc()),
    )


class NewsArticleSymbol(Base):
    """Many-to-many link between an article and the instrument(s) it covers.

    ``symbol`` is the internal code (``AAPL.US``, ``TSLA.US``, ...).
    ``match_type`` records whether the ticker was found in the title,
    body, cashtag, subreddit name, etc. — useful for tuning extraction.
    """

    __tablename__ = "news_article_symbol"

    article_id = Column(
        BigInteger,
        ForeignKey("news_article.id", ondelete="CASCADE"),
        primary_key=True,
    )
    symbol = Column(String(20), primary_key=True, comment="Internal code, e.g. AAPL.US")
    match_type = Column(String(30), comment="title | body | cashtag | subreddit")
    confidence = Column(Integer, default=100, comment="0..100 extraction confidence")

    article = relationship("NewsArticle", back_populates="symbols")

    __table_args__ = (
        Index("ix_news_article_symbol_symbol", "symbol"),
    )


class RedditCommentCache(Base):
    """Cache of Reddit comments (second-level discussion)."""

    __tablename__ = "reddit_comment_cache"

    reddit_id = Column(String(20), primary_key=True, comment="Reddit comment id (t1_xxx)")
    parent_id = Column(String(20), index=True, comment="Parent post or comment id")
    subreddit = Column(String(50), index=True)
    author = Column(String(100))
    body = Column(Text)
    score = Column(Integer, default=0)
    controversiality = Column(Integer, default=0)
    created_utc = Column(DateTime)
    fetched_at = Column(DateTime, server_default=func.now(), nullable=False)
